File: ios/SWIFT/Beacon/3-Alert/Complete/AlertStrategy/TimeConditionsFirstWay.swift
//
//  TimeConditionsFirstWay.swift
//  AlertStrategy
//
//  Created by Ludovic Vimont on 13/11/2017.
//  Copyright © 2017 R&D connecthings. All rights reserved.
//

import Foundation
import ConnectPlaceCommon
import ConnectPlaceActions

public class TimeConditionsFirstWay: InAppActionConditions, InAppActionConditionsUpdater {
    let maxTimeBeforeReset: Double
    let timeProvider: TimeProvider
    let delayBeforeCreation: Double

    public init(maxTimeBeforeReset: Double = 60, delayBeforeCreation: Double = 60) {
        self.maxTimeBeforeReset = maxTimeBeforeReset
        self.delayBeforeCreation = delayBeforeCreation
        self.timeProvider = TimeProviderAbsolute()
    }
    
    public func getApplicationNamespace() -> String {
        return String(reflecting: TimeConditionsFirstWay.self).split(separator: ".")[0].description
    }
    
    public func getConditionsKey() -> String {
        return "inAppActionTimeConditions"
    }
    
    public func getInAppActionParameterClass() -> String {
        return "TimeConditionsParameter"
    }
    
    public func updateInAppActionConditionsParameter(_ conditionsParameter: InAppActionConditionsParameter) {
        updateLastDetectionTime(conditionsParameter)
        conditionsParameter.isConditionValid = isConditionValid(conditionsParameter)
        conditionsParameter.inAppActionStatus = updateInAppActionStatus(conditionsParameter)
        conditionsParameter.isReadyForAction = isReadyForAction(conditionsParameter)
        conditionsParameter.maxTimeBeforeActionDoneReset = updateMaxTimeBeforeActionDoneReset(conditionsParameter)
    }

    public func updateLastDetectionTime(_ conditionsParameter: InAppActionConditionsParameter) {
        if let strategyParameter = conditionsParameter as? TimeConditionsParameter {
            let currentTime: Double = timeProvider.getTime()
            if strategyParameter.lastDetectionTime + 3 < currentTime {
                GlobalLogger.shared.debug("delay before creating " , delayBeforeCreation)
                strategyParameter.timeToShowAlert = currentTime + delayBeforeCreation
            }
            strategyParameter.lastDetectionTime = currentTime + 0
        }
    }

    public func isConditionValid(_ conditionsParameter: InAppActionConditionsParameter) -> Bool {
        if let strategyParameter = conditionsParameter as? TimeConditionsParameter {
            GlobalLogger.shared.debug("remaining time ",(timeProvider.getTime() - strategyParameter.timeToShowAlert))
            return strategyParameter.timeToShowAlert < timeProvider.getTime()
        } else {
            return true
        }
    }

    public func isReadyForAction(_ conditionsParameter: InAppActionConditionsParameter) -> Bool {
       return conditionsParameter.inAppActionStatus == InAppActionStatus.WAITING_FOR_ACTION
            && conditionsParameter.isConditionValid
    }

    public func updateMaxTimeBeforeActionDoneReset(_ conditionsParameter: InAppActionConditionsParameter) -> Double {
        return timeProvider.getTime() + maxTimeBeforeReset
    }

    public func updateInAppActionStatus(_ conditionsParameter: InAppActionConditionsParameter) -> InAppActionStatus {
        if conditionsParameter.inAppActionStatus == InAppActionStatus.DONE {
            if !conditionsParameter.isConditionValid
                || conditionsParameter.maxTimeBeforeActionDoneReset < timeProvider.getTime() {
                return InAppActionStatus.WAITING_FOR_ACTION
            }
        }
        return conditionsParameter.inAppActionStatus
    }
}

File: android/beacon/3-Alert/Complete/app/src/main/java/android/connecthings/com/tuto/alert/TimeConditionsFirstWay.java
package android.connecthings.com.tuto.alert;

import android.os.SystemClock;

import com.connecthings.connectplace.actions.inappaction.conditions.InAppActionConditions;
import com.connecthings.connectplace.actions.inappaction.conditions.InAppActionConditionsUpdater;
import com.connecthings.connectplace.actions.inappaction.conditions.parameter.InAppActionConditionsParameter;
import com.connecthings.connectplace.actions.inappaction.enums.InAppActionActionStatus;
import com.connecthings.connectplace.actions.inappaction.updater.InAppActionManagerUpdater;
import com.connecthings.util.Log;

/**
 * Created by Connecthings on 14/11/2017.
 */
public class TimeConditionsFirstWay implements InAppActionConditions, InAppActionConditionsUpdater {
    private int maxTimeBeforeReset;
    private double delayBeforeCreation;

    public TimeConditionsFirstWay(int maxTimeBeforeReset, double delayBeforeCreation) {
        this.maxTimeBeforeReset = maxTimeBeforeReset;
        this.delayBeforeCreation = delayBeforeCreation;
    }

    @Override
    public String getConditionsKey() {
        return "timeConditions";
    }

    @Override
    public Class<? extends InAppActionConditionsParameter> getInAppActionParameterClass() {
        return TimeConditionsParameter.class;
    }

    @Override
    public void updateInAppActionConditionsParameter(InAppActionConditionsParameter inAppActionConditionsParameter) {
        updateTimeConditionsParameter(inAppActionConditionsParameter);
        inAppActionConditionsParameter.setConditionValid(isConditionValid(inAppActionConditionsParameter));
        inAppActionConditionsParameter.setInAppActionActionStatus(updateInAppActionActionStatus(inAppActionConditionsParameter));
        inAppActionConditionsParameter.setReadyForAction(isReadyForAction(inAppActionConditionsParameter));
        if (inAppActionConditionsParameter.getActionStatus() != InAppActionActionStatus.DONE) {
            inAppActionConditionsParameter.setMaxTimeBeforeActionDoneReset(updateMaxTimeBeforeActionDoneReset(inAppActionConditionsParameter));
        }
    }

    public void updateTimeConditionsParameter(InAppActionConditionsParameter inAppActionConditionsParameter) {
        TimeConditionsParameter timeConditionsParameter = (TimeConditionsParameter) inAppActionConditionsParameter;
        long currentTime = SystemClock.elapsedRealtime();
        if (timeConditionsParameter.getLastDetectionTime() + 3000 < SystemClock.elapsedRealtime()) {
            timeConditionsParameter.setTimeToShowAlert(currentTime + delayBeforeCreation);
        }
        timeConditionsParameter.setLastDetectionTime(currentTime);
    }

    @Override
    public boolean isConditionValid(InAppActionConditionsParameter inAppActionConditionsParameter) {
        TimeConditionsParameter timeConditionsParameter = (TimeConditionsParameter) inAppActionConditionsParameter;
        Log.d("TimeConditions", "remaining time before alert: ", (timeConditionsParameter.getTimeToShowAlert() - SystemClock.elapsedRealtime()));
        return timeConditionsParameter.getTimeToShowAlert() < SystemClock.elapsedRealtime();
    }

    @Override
    public boolean isReadyForAction(InAppActionConditionsParameter inAppActionConditionsParameter) {
        return inAppActionConditionsParameter.getActionStatus() == InAppActionActionStatus.WAITING_FOR_ACTION && inAppActionConditionsParameter.isConditionValid();
    }

    @Override
    public long updateMaxTimeBeforeActionDoneReset(InAppActionConditionsParameter inAppActionConditionsParameter) {
        return SystemClock.elapsedRealtime() + maxTimeBeforeReset;
    }

    @Override
    public InAppActionActionStatus updateInAppActionActionStatus(InAppActionConditionsParameter inAppActionConditionsParameter) {
        if (inAppActionConditionsParameter.getActionStatus() == InAppActionActionStatus.DONE
                && (!inAppActionConditionsParameter.isConditionValid())
                || inAppActionConditionsParameter.getMaxTimeBeforeActionDoneReset() < SystemClock.elapsedRealtime()) {
            return InAppActionActionStatus.WAITING_FOR_ACTION;
        }
        return inAppActionConditionsParameter.getActionStatus();
    }
}